core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java

```diff
@@ -217,7 +226,13 @@ public PartitionSpec apply() {
         if (newName != null) {
           builder.add(field.sourceId(), field.fieldId(), newName, field.transform());
         } else {
-          builder.add(field.sourceId(), field.fieldId(), field.name(), field.transform());
+          // field name of void transform will be replaced by field name append with UUID if conflict is detected.
+          String partitionName = field.name();
+          if (isVoidTransform(field) && (renames.containsValue(partitionName) ||
```

Comment:
The logic here is still not quite what I was picturing, although I think this is a bit cleaner than before.

What I was thinking is that we can detect name collisions in `addField` by checking for an existing field with some name. If one exists, then instead of throwing an exception we could check if it is a void transform and add a rename if it is. Something like this:

```java
  public BaseUpdatePartitionSpec addField(String name, Term term) {
    PartitionField existingField = nameToField.get(name);
    if (existingField != null) {
      if (isVoidTransform(existingField)) {
        renameField(name, name + "_" + existingField.fieldId());
      } else {
        throw new IllegalArgumentException(String.format("Cannot add duplicate partition field name: %s", name));
      }
    }
    ...
  }
```

Would that not work?

Reply:
This is better than my code. Already move detect code to `addField` and `renameField` in newest commit.